### PR TITLE
Test cleanup

### PR DESCRIPTION
**Pull request by genio**
Cleanup the entire test suite to use strict, warnings, and Test::More.  It seems to be testing well in 6.15_004

### CODE REVIEW

**Comment by genio** [comment]
This was a pretty huge PR to do in hindsight. I don't think anyone has the time to go through this in its entirety to see if it's ok to merge (I wouldn't want to be the one to pull the trigger on this).

👎 

Should I break this up in a way that's a bit easier to go through and audit (one test file at a time, etc)?  Or, is this entire idea really unnecessary in the long run?  I don't want to be the creator of clutter. :)

**Comment by genio** [comment]
Closing this out.

**Comment by genio** [comment]
@oalders If there are things you're less than thrilled about, please let me know and I'll be happy to rework in anyway you see fit.  If it needs to be broken up and reviewed in smaller bits, we can get that done as well.

File: t/base/protocols.t [hunk 1]
```diff
@@ -1,20 +1,23 @@
 use strict;
 use warnings;
 use Test::More;
-plan tests => 6;
 
 use LWP::UserAgent;
+plan tests => 8;
```
Comment (oalders) [inline t/base/protocols.t]: I think, in general, we no longer need to a plan.  `done_testing()` at the end of the file is fine.

File: t/base/protocols.t [hunk 2]
```diff
@@ -1,20 +1,23 @@
...
 
 my $ua = LWP::UserAgent->new();
+isa_ok($ua, 'LWP::UserAgent', 'new: UserAgent instance');
```
Comment (oalders) [inline t/base/protocols.t]: I'm also not sure how helpful these `isa_ok` checks are for `LWP::UserAgent`.  I can see it for a method that is supposed to return some kind of an object, but I'm not sure we're learning anything useful here.

File: t/base/proxy.t [hunk 3]
```diff
@@ -1,17 +1,20 @@
...
     my $ua = LWP::UserAgent->new;
-    eval { $ua->env_proxy(); };
-    ok($@, "");
+    isa_ok($ua, 'LWP::UserAgent', 'new: UserAgent instance');
+    my $error = do {
+        local $@;
+        eval { $ua->env_proxy(); };
+        $@;
+    };
+    is($error, "", "proxy: with env: $varname: no errors");
```
Comment (oalders) [inline t/base/proxy.t]: This could be cleaned up a bit by using `Test::Fatal` rather than the `eval`.

File: t/live/apache-http10.t [hunk 4]
```diff
@@ -1,16 +1,21 @@
...
+
 $req->protocol("HTTP/1.0");
+
 my $res = $ua->simple_request($req);
-ok($res->content, qr/HTTP\/1.0/);
+isa_ok($res, 'HTTP::Response', 'simple_request: Got a proper response');
+like($res->content, qr/HTTP\/1.0/, 'Request to apache.org: Got an HTTP 1.0 response');
 
-$res->dump(prefix => "# ");
+# $res->dump(prefix => "# ");
```
Comment (oalders) [inline t/live/apache-http10.t]: If it's going to be commented out, we can probably just remove it entirely.

File: t/live/jigsaw/chunk.t [hunk 5]
```diff
@@ -1,37 +1,35 @@
...
+use HTTP::Request;
 use LWP::UserAgent;
 
+plan tests => 10;
+
 my $ua = LWP::UserAgent->new(keep_alive => 1);
+isa_ok($ua, 'LWP::UserAgent', 'new: UserAgent instance');
 
 my $req = HTTP::Request->new(GET => "http://jigsaw.w3.org/HTTP/ChunkedScript");
+isa_ok($req, 'HTTP::Request', 'new: HTTP::Request instance');
```
Comment (oalders) [inline t/live/jigsaw/chunk.t]: I realize this isn't your fault, but if there are tests that need the network, we should be using something like `Test::RequiresInternet`.

File: t/live/jigsaw/md5-get.t [hunk 6]
```diff
@@ -1,29 +1,29 @@
...
 use LWP::UserAgent;
 
+plan tests => 6;
+
 my $ua = LWP::UserAgent->new(keep_alive => 1);
+isa_ok($ua, 'LWP::UserAgent', 'new: UserAgent instance');
 
 my $res = $ua->get(
-  "http://jigsaw.w3.org/HTTP/h-content-md5.html",
+    "http://jigsaw.w3.org/HTTP/h-content-md5.html",
```
Comment (oalders) [inline t/live/jigsaw/md5-get.t]: Again, `Test::RequiresInternet`.

File: t/live/jigsaw/te.t [hunk 7]
```diff
@@ -1,33 +1,31 @@
...
     }
-    print $req->as_string;
 
     my $res = $ua->request($req);
+    isa_ok($res, 'HTTP::Response', 'request: Got a proper response');
     if (defined $content) {
-	print "not " unless $content eq $res->content;
-	print "ok $testno\n\n";
-	$testno++;
+        is($res->content, $content, 'content: Correct content');
```
Comment (oalders) [inline t/live/jigsaw/te.t]: 👍

File: t/local/get.t [hunk 8]
```diff
@@ -1,87 +1,75 @@
...
-    if (open(TEST, ">$_/test-$$")) {
-        close(TEST);
-	unlink("$_/test-$$");
-	$TMPDIR = $_;
-	last;
-    }
+my $TMPDIR = undef;
+
+if ( $^O eq 'MacOS' ) {
+    plan skip_all => 'Cannot test on this platform';
```
Comment (oalders) [inline t/local/get.t]: I wonder what the reason for this was?  I bet it's fine on MacOS now.

**Comment by oalders** [comment]
Also, we'll need to get this running on Travis.  We should probably have the live tests running there as well.

File: t/base/protocols.t (hunk 1 above)
Comment (genio) [inline t/base/protocols.t]: I tend to agree with you.  However, after some discussion in #lwp this is preferable to requiring at least version 0.88 of Test::More.

I'll be happy to update things if we're willing to make that concession.

File: t/local/get.t (hunk 8 above)
Comment (genio) [inline t/local/get.t]: MacOS at this point is identified as 'darwin'.  MacOS is for Mac OS 7, 8, and 9.  I don't know the reasoning behind not allowing the tests on older macs, but maybe this is no longer necessary.

File: t/base/protocols.t (hunk 1 above)
Comment (oalders) [inline t/base/protocols.t]: Do you remember what the reasoning behind this was?

File: t/base/proxy.t (hunk 3 above)
Comment (genio) [inline t/base/proxy.t]: Done.

File: t/live/apache-http10.t (hunk 4 above)
Comment (genio) [inline t/live/apache-http10.t]: Done

**Comment by genio** [comment]
Travis is now passing.  YAY

**Comment by shadowcat-mst** [comment]
Sprinkling isa_ok after constructors is silly. All you're doing there really is testing 'new() is not completely broken'.

**Comment by genio** [comment]
I think that all of the things this initial review have brought up are now corrected. Are there any feelings about whether this mass move to Test::More is worth it?

Please keep in mind that this initial effort was solely to get things using strict/warnings and Test::More. It's only meant as a first-pass attempt to get the test suite somewhat modernized. If I've gone about it in a bad/wrong/different way than what you'd like to do, we can nix it all together. No harm, no foul.

On the other hand, if this is moving in a positive direction but is just too big to feel comfortable with, we can break this somewhat cleaned up and reviewed PR up into smaller bits so a more detailed code review can be done.

In other words, I want to help, but my feelings won't be hurt if this particular set of help isn't very helpy. I'm free to discuss it on IRC, just yell at me.

**Comment by shadowcat-mst** [comment]
Having to use no_plan rather than done_testing to avoid the 0.88 dep is, I'm sure, faintly annoying, but (1) we might as well (2) I approve of the general principle of this.

File: t/base/protocols.t (hunk 1 above)
Comment (oalders) [inline t/base/protocols.t]: From #perl-qa:

```
[11:10:31]  <@oalders>	What would be a compelling reason for not requiring a Test::More greater than 0.88? https://github.com/libwww-perl/libwww-perl/pull/88#discussion_r81676260
[11:10:33]  <+dipsy>	[ Test cleanup by genio · Pull Request #88 · libwww-perl/libwww-perl · GitHub ] 
[11:11:54]  <ilmari>	it wasn't core until 5.10.1, so you'd force 5.8 users to upgrade
[11:13:23]  <@oalders>	ah, ok.  is there any consensus on how many 5.8 users there actually are?  i see it come up every so often
[11:15:18]  <@mst>	0.88 is just for done_testing rather than no_plan
[11:15:21]  <@mst>	which is barely a big deal
```

I think it's fine to ignore my `done_testing()` suggestion. :)

**Comment by oalders** [comment]
I'm on board with this generally as well.  I'll have another pass at checking the latest changes this week, but I think we should merge the whole thing if we don't find any problems.

**Comment by genio** [comment]
@oalders Let me know if this massive thing could be re-organized in a way that would make reviewing easier on you.  Directory by directory, file by file, etc.  Also, would you prefer a squash and force push to make it one single commit rather than the many above? etc.

Basically, whatever I can do to make it easier...

File: t/local/autoload.t [hunk 9]
```diff
@@ -1,22 +1,32 @@
...
 
 my $request = HTTP::Request->new(GET => $url);
 
 my $response = $ua->request($request);
-ok($response->is_success);
+isa_ok($response, 'HTTP::Response', 'Got a proper response');
+
+ok( $response->is_success(), 'Response was successful' );
+unless($response->is_success()) {
+    print $response->error_as_HTML;
```
Comment (oalders) [inline t/local/autoload.t]: Should this be a `diag`?

File: t/local/http.t [hunk 10]
```diff
@@ -1,498 +1,475 @@
...
+    }
+    { # put
+        my $res = $ua->put(url("/echo/path_info?query", $base),
+            Accept => 'text/html',
+            Accept => 'text/plain; q=0.9',
+            Accept => 'image/*',
+            X_Foo => "Bar",
+        );
+        isa_ok($res, 'HTTP::Response', 'put: good response object');
+        #$res->dump;
```
Comment (oalders) [inline t/local/http.t]: We can delete this line.

File: t/local/http.t [hunk 11]
```diff
@@ -1,498 +1,475 @@
...
+    }
+    { # delete
+        my $res = $ua->delete(url("/echo/path_info?query", $base),
+            Accept => 'text/html',
+            Accept => 'text/plain; q=0.9',
+            Accept => 'image/*',
+            X_Foo => "Bar",
+        );
+        isa_ok($res, 'HTTP::Response', 'delete: good response object');
+        #$res->dump;
```
Comment (oalders) [inline t/local/http.t]: We can delete this line as well.

File: t/local/http.t [hunk 12]
```diff
@@ -1,498 +1,475 @@
...
+    $router{delete_echo} = sub {
+        my($c, $req) = @_;
+        $c->send_basic_header(200);
+        $c->print("Content-Type: message/http\015\012");
+        $c->send_crlf;
+        $c->print($req->as_string);
+    };
+    $router{get_basic} = sub {
+        my($c, $r) = @_;
+        #print STDERR $r->as_string;
```
Comment (oalders) [inline t/local/http.t]: We can delete this line.

File: t/local/http.t [hunk 13]
```diff
@@ -1,498 +1,475 @@
...
+        }
+        else {
+            $c->send_basic_header(401);
+            $c->print("WWW-Authenticate: Basic realm=\"libwww-perl\"\015\012");
+            $c->send_crlf;
+        }
+    };
+    $router{get_digest} = sub {
+        my($c, $r) = @_;
+        # print STDERR $r->as_string;
```
Comment (oalders) [inline t/local/http.t]: We can delete this line.

File: t/local/http.t [hunk 14]
```diff
@@ -1,498 +1,475 @@
...
+        if ( defined($auth) && $auth =~ /^Digest\s+(.*)$/ ) {
+            %auth_params = map { split /=/ } split /,\s*/, $1;
+        }
+        if ( %auth_params &&
+                $auth_params{username} eq "\"ok 23\"" &&
+                $auth_params{realm} eq "\"libwww-perl-digest\"" &&
+                $auth_params{qop} eq "auth" &&
+                $auth_params{algorithm} eq "\"MD5\"" &&
+                $auth_params{uri} eq "\"/digest\"" &&
+                $auth_params{nonce} eq "\"12345\"" &&
```
Comment (oalders) [inline t/local/http.t]: Could we make the quotes more readable?

`$auth_params{nonce} eq q{"12345"} &&`

File: t/local/protosub.t [hunk 15]
```diff
@@ -1,70 +1,58 @@
...
 
-sub request
-{
-    my $self = shift;
-    my($request, $proxy, $arg, $size, $timeout) = @_;
-    #print $request->as_string;
+    sub request {
+        my $self = shift;
+        my($request, $proxy, $arg, $size, $timeout) = @_;
+        #print $request->as_string;
```
Comment (oalders) [inline t/local/protosub.t]: We can delete this line.

File: t/local/autoload.t (hunk 9 above)
Comment (genio) [inline t/local/autoload.t]: done

**Comment by genio** [comment]
@oalders The last set of requested changes are now complete.

**Comment by oalders** [comment]
\o/
